Refactor Engine, fixes #256

trishume opened the pull request:
This PR contains a number of related refactorings to the `Engine`, `Delta` and `Subset` classes. It is entirely re-namings, doc comment changes, and one removal of a duplicate method. Split into multiple commits according to which refactors must occur together, you may want to review each commit individually. Fixes #256, see discussion there for context.

cc @raphlinus 

I've pasted the commit message of the most significant refactoring below:

### Rename methods of Subset to be more consistent.

There was previously an inconsistency in the naming in Engine and the naming
in Subset. See #256

The naming of the methods of Subset was as if the Subset represented
the complement of the ranges it contained, which if the ranges were deletions
meant it represented the elements that were kept.

However, this was inconsistent with the naming in Engine which had subsets like
`inserts` and `deletes` that according to the concept of Subset actually represented
the non-inserted and non-deleted ranges.

This fixes the inconsistency by renaming methods as if Subset just represents the
ranges it contains, which eliminates having to reason about the complement when
thinking about the entire system.

cmyr commented on rust/rope/src/delta.rs:
```diff
@@ -22,10 +22,16 @@ use subset::{Subset, SubsetBuilder};
 use std::cmp::min;
 
 enum DeltaElement<N: NodeInfo> {
+    /// Represents the text in a range of the base document. Includes beginning, excludes end.
```
maybe "a range in the base document", or "a range of text in the base document".

cmyr commented on rust/rope/src/delta.rs:
```diff
@@ -22,10 +22,16 @@ use subset::{Subset, SubsetBuilder};
     Insert(Node<N>),
 }
 
+/// Represents an entire new document as a sequence of sections copied from
```
I find this a little confusing. I suppose it is a representation of a document, but I find it easier to think of it as a representation of a _transformation_ from one document (or document state) to another. 

This PR is a welcome opportunity for me to clarify my own understanding of a bunch of this code. 

How I would write this:

"A representation of a transformation from one document state to another. 

This transformation is composed of a series of `Copy` and `Insert` operations. `Copy` operations indicate unmodified ranges of the existing document, and `Insert` operations indicate additions. 

Note: The deletion of some part of the document, in this schema, is represented as a set of `Copy` operations over the ranges to be kept.

For example, the transformation from "hello" to "hippo" can be represented as:
`[Copy(0, 1), Insert("ipp"), Copy(4,5)]`"



(sorry, I got a little carried away)

trishume replied:
I agree to some extent, but the reason I changed it is that while the struct conceptually represents a change, something that confused me at first was that the `els` field actually represents the entire new document, just by copying things from the old one.

I've changed the comment to reflect both of these things better.

cmyr commented on rust/rope/src/delta.rs:
```diff
@@ -205,10 +211,10 @@ impl<N: NodeInfo> Delta<N> {
     }
 
-    /// Return a subset that inverts the insert-only delta:
+    /// Return a Subset containing the inserted ranges of an insert-only delta.
```
Something that isn't clear to me: is "insert-only" a guarantee made anywhere? Is the caller expected to ensure that a given delta is an "insert-only delta"?

cmyr commented on rust/rope/src/engine.rs:
```diff
@@ -85,28 +85,22 @@ impl Engine {
     }
 
+    /// Get the contents of the document at a given revision number
     fn get_rev_from_index(&self, rev_index: usize) -> Rope {
```
I think this function might be another candidate for a renaming. Maybe something like `get_revision`? or `document_at_revision`? I think the word `from` is confusing, given the way it is used in Rust to indicate conversion constructors. Maybe `for` would be more communicative? I also don't love `get` for something that isn't a getter. 😕

cmyr commented on rust/rope/src/subset.rs:
```diff
@@ -48,57 +48,54 @@ impl SubsetBuilder {
     // Maybe Subset should be a pure data structure and this method should
     // be a method of Node.
-    pub fn apply<N: NodeInfo>(&self, s: &Node<N>) -> Node<N> {
+    pub fn delete_in<N: NodeInfo>(&self, s: &Node<N>) -> Node<N> {
```
This would benefit from documentation.

I also prefer `delete_from` to `delete_in`.

raphlinus commented on rust/rope/src/delta.rs:
```diff
@@ -145,27 +159,65 @@ impl<N: NodeInfo> Delta<N> {
+        (Interval::new_closed_open(iv_start, iv_end), Delta::els_len(els))
+    }
+
+    fn els_len(els: &[DeltaElement<N>]) -> usize {
```
Doc (and possible rename): this method returns the length of the transformed string.

raphlinus commented on rust/rope/src/engine.rs:
```diff
@@ -33,7 +33,7 @@ pub struct Engine {
 struct Revision {
     rev_id: usize,
-    from_union: Subset,
+    deletes_from_union: Subset,
```
yes, this is a good name. probably main reason i didn't go there is i tend to favor terseness

trishume replied:
It was a totally fine name back when `Subset` actually was supposed to represent a transformation of a document to a subset of it, then the type told you what it was. Under the new model though, it needs a longer name to be unambiguous.

Also, Sublime has really good auto-complete for terms that are used frequently within a single file, so for me it isn't any more difficult to type. Everyone else should get a better editor, although I know they're working on it. 😉

raphlinus commented on rust/rope/src/subset.rs:
```diff
@@ -48,57 +48,55 @@ impl SubsetBuilder {
-    /// resulting subset iff it exists in both inputs.
-    pub fn intersect(&self, other: &Subset) -> Subset {
+    /// Compute the union of two subsets. In other words, an element exists in the
+    /// resulting subset if it exists either input.
```
'iff' is correct here, but maybe people aren't familiar with the usage?

trishume replied:
Oh that was just an accidental omission when I edited it to describe union instead of intersection. I just improved it.

trishume commented:
@cmyr I addressed all your comments, thanks, they were valuable and I agreed with them all.

This PR now has a lot of commits, but before merging I plan on doing some squashing, although I don't want to squash the entire thing since it's useful to pick out the individual changes from being washed away in the diffs of the larger ones.
